src/utils/SparseGrid.h

```diff
@@ -0,0 +1,170 @@
+/** Copyright (C) 2016 Scott Lenser - Released under terms of the AGPLv3 License */
+
+#ifndef SPARSE_GRID_H
+#define SPARSE_GRID_H
+
+#include "intpoint.h"
+
+#include <unordered_map>
+#include <vector>
+
+namespace cura {
+
+/*! \brief Sparse grid which can locate spatially nearby values efficiently.
+ *
+ * \tparam Val The value type to store.
+ * \tparam PointAccess The functor to get the location from Val.  PointAccess
+ *    must have: Point operator()(const Val &val) const
+ *    which returns the location associated with val.
+ */
+template<class Val, class PointAccess>
+class SparseGrid
+{
+public:
+    /*! \brief Constructs a sparse grid with the specified cell size.
+     *
+     * \param[in] cell_size The size to use for a cell (square) in the grid.
+     *    Typical values would be around 0.5-2x of expected query radius.
+     */
+    SparseGrid(coord_t cell_size);
+
+    /*! \brief Inserts val into the sparse grid.
+     *
+     * \param[in] val The value to be inserted.
+     */
+    void insert(const Val &val);
+
+    /*! \brief Returns all data within radius of query_pt.
+     *
+     * Finds all values with location within radius of \p query_pt.  May
+     * return additional values that are beyond radius.
+     *
+     * Average running time is a*(1 + 2 * radius / cell_size)**2 +
+     * b*cnt where a and b are proportionality constance and cnt is
+     * the number of returned items.  The search will return items in
+     * an area of (2*radius + cell_size)**2 on average.  The max range
+     * of an item from the query_point is radius + cell_size.
+     *
+     * \param[in] query_pt The point to search around.
+     * \param[in] radius The search radius.
+     * \return Vector of values found
+     */
+    std::vector<Val> getNearby(const Point &query_pt, coord_t radius) const;
+
+private:
+    using GridPoint = Point;
+    using GridMap = std::unordered_multimap<GridPoint, Val>;
+
+    /*! \brief Add values from the cell indicated by \p grid_pt.
+     *
+     * \param[in] grid_pt The grid coordinates of the cell.
+     * \param[out] ret The values in the cell are appended to ret.
+     */
+    void addFromCell(const GridPoint &grid_pt,
+                     std::vector<Val> &ret) const;
+
+    /*! \brief Add values from cells that might contain sought after points.
+     *
+     * Appends values from cell that might have values within \p
+     * radius of \p query_pt.  Appends all values that are within
+     * radius of query_pt.  May append values that are up to radius +
+     * cell_size from query_pt.
+     *
+     * \param[in] query_pt The point to search around.
+     * \param[in] radisu The search radius.
+     * \param[out] ret The values in the cell are appended to ret.
+     */
+    void addNearby(const Point &query_pt, coord_t radius,
```

Comment:
Oh dear, I didn't know that line would be interpreted that way. That's horrible!

Perhaps we should forbid one-line multiple declarations.

Anyway the code base buts the ampersand and start with the type, so it would be nice to keep that consistent. If you don't change it I will...